runtime/parachains/src/session_info.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The session info module provides information about validator sets
+//! from prior sessions needed for approvals and disputes.
+//!
+//! See https://w3f.github.io/parachain-implementers-guide/runtime/session_info.html.
+
+use primitives::v1::{SessionIndex, SessionInfo, ValidatorId};
+use frame_support::{
+	decl_storage, decl_module, decl_error,
+	weights::Weight,
+};
+use crate::{configuration, paras, scheduler};
+
+pub trait Trait:
+	frame_system::Trait
+	+ configuration::Trait
+	+ paras::Trait
+	+ scheduler::Trait
+	+ pallet_authority_discovery::Trait
+{
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as ParaSessionInfo {
+		/// The earliest session for which previous session info is stored.
+		EarliestStoredSession get(fn earliest_stored_session): SessionIndex;
+		/// Session information in a rolling window.
+		/// Should have an entry in range `EarliestStoredSession..=CurrentSessionIndex`.
+		Sessions get(fn session_info): map hasher(identity) SessionIndex => Option<SessionInfo>;
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> { }
+}
+
+decl_module! {
+	/// The session info module.
+	pub struct Module<T: Trait> for enum Call where origin: <T as frame_system::Trait>::Origin {
+		type Error = Error<T>;
+	}
+}
+
+impl<T: Trait> sp_runtime::BoundToRuntimeAppPublic for Module<T> {
+	type Public = ValidatorId;
+}
+
+impl<T: Trait> Module<T> {
+	/// Handle an incoming session change.
+	pub(crate) fn initializer_on_new_session(
+		notification: &crate::initializer::SessionChangeNotification<T::BlockNumber>
+	) {
+		let config = <configuration::Module<T>>::config();
+
+		let dispute_period = config.dispute_period;
+		let n_parachains = <paras::Module<T>>::parachains().len() as u32;
+
+		let validators: Vec<_> = notification.validators.clone();
+		let discovery_keys = <pallet_authority_discovery::Module<T>>::authorities();
```

Comment:
is `pallet_authority_discovery` guaranteed to update it's storage `on_new_session` before that call?

Reply:
cc @mxinden 

We delay session change notifications to parachains pallets by 1 block, so I believe that the authority discovery module, which doesn't have this delay, will have been updated at the prior block.